File: single_linked_list.py
class Node:
	def __init__(self, value):
		self.value = value
		self.next = None
	def getValue(self):
		return self.value
	def getNext(self):
		return self.next
	def setValue(self, value):
		self.value = value
	def setNext(self, next):
		self.next = next

#node_one = Node(1)
#print (node_one.getValue())
#print (node_one.value)

class SingleLinkedList:
	def __init__(self):
		self.head = None

	def isEmpty(self):
		return self.head == None

	def addFront(self,value):
		nextNode = Node(value)
		if self.isEmpty():
			self.head = nextNode
		else:
			nextNode.next = self.head
			self.head = nextNode

	def addBack(self, value):
		newNode = Node(value)
		current = self.head
		if not current:
			self.head = newNode
			return
		while current.getNext():
			current = current.getNext()
		current.next = newNode

	def size(self):
		if self.isEmpty():
			print ("0")
			return 0
		count = 0
		current = self.head
		while current:
			count += 1
			current = current.getNext()
		print (count)
		return count

	#return false if not in list, else return index
	def search(self, value):
		if self.isEmpty():
			return False
		current = self.head
		index = 0
		while current:
			if current.getValue() == value:
				print ("Found " + str(value) + " at index: " + str(index))
				return index
			else:
				index += 1
				current = current.getNext()
		print ("couldn't find it")
		return False

	def removeAtIndex(self, targetIndex):
		if self.isEmpty() or targetIndex > self.size()-1:
			return False
		if targetIndex == 0:
			self.head = self.head.getNext()
			return
		current = self.head
		previous = current
		index = 0
		while index < targetIndex:
			index += 1
			previous = current
			current = current.getNext()
		print ("should only print 2 numbers")
		print (previous.getValue())
		print (current.getValue())
		previous.setNext(current.getNext())
		current.setNext(None)

	def removeValue(self, targetValue):
		index = self.search(targetValue)
		if not index:
			return False
		self.removeAtIndex(index)
		return True



	def printList(self):
		if not self.head:
			print ("[]")
			return
		listStr = "["
		current = self.head
		while current:
			listStr += str(current.value)
			if current.next:
				listStr += ", "
			current = current.getNext()
		listStr += "]"
		print (listStr)

mySLL = SingleLinkedList()
mySLL.size()
mySLL.addBack(2)
mySLL.addFront(1)
mySLL.printList()
mySLL.addFront(0)
mySLL.printList()
mySLL.addBack(3)
mySLL.printList()
mySLL.size()
mySLL.search(999)
mySLL.search(2)
mySLL.printList()

print ("((((((((((((((((((((")
mySLL.removeAtIndex(0)
mySLL.addBack(4)
mySLL.addBack(5)
mySLL.printList()
mySLL.removeValue(3)
mySLL.printList()

File: singleLinkedList.js
class Node {
    constructor(value) {
        this.value = value;
        this.next = null;
    }

    getValue() {
        return this.value;
    }

    getNext() {
        return this.next;   
    }

    setValue(value) {
        this.value = value;
    }
    
    setNext(next) {
        this.next = next;
    }
}

class SingleLinkedList {
    constructor() {
        this.head = null;
    }

    isEmpty() {
        return this.head === null;
    }

    addFront(value){
        let newNode = new Node(value);
        if (this.isEmpty()) {
            this.head = newNode;
            return this;
        } else {
            newNode.setNext(this.head);
            this.head = newNode;
            return this;
        }       
    }

    printList(value){
        let strList = "["
        let current = this.head
        while (current) {
            strList += current.getValue();
            if (current.getNext()){
                strList += ", "
            }
            current = current.getNext();
        }
        strList += "]"
        console.log(strList);
        return this;
    }

    addBack(value) {
        let newNode = new Node(value);
        if (this.isEmpty()) {
            this.head = newNode;
            return this;
        }
        let current = this.head;
        while (current.getNext()) {
            current = current.getNext();
        }
        current.setNext(newNode);
        return this;
    }

    size(){
        if (this.isEmpty()){
            console.log("The size of the list is: " + 0);
            return 0;
        }
        let count = 0;
        let current = this.head;
        while (current) {
            count++;
            current = current.getNext();
        }
        console.log("The size of the list is: " + count);
        return count;
    }

    findValue(targetValue){
        if (this.isEmpty()){
            return false;
        }
        let index = 0;
        let current = this.head;
        while (current){
            if (current.getValue() == targetValue){
                console.log("found ya")
                return index;
            }
            index++;
            current = current.getNext();
        }
        return false;
    }

    removeAtIndex(targetIndex){
        if (this.isEmpty() || targetIndex > this.size()-1) {
            return false;
        }
        if (targetIndex == 0){
            this.head = this.head.next;
            this.printList();
            return true;
        }

        let current = this.head;
        let previous = current;
        let index = 0;
        while (index < targetIndex){
            previous = current;
            current = current.getNext();
            index++;
        }
        previous.setNext(current.getNext());
        current.setNext(null);
        return true;
    }

    removeValue(targetValue){
        if (this.isEmpty()|| !this.findValue(targetValue)){
            return false;
        }
        return this.removeAtIndex(this.findValue(targetValue));
    }
}


const myList = new SingleLinkedList();
myList.printList();
myList.size();

myList.addFront(0);
myList.printList();
myList.size();

myList.addBack(3);
myList.printList();
myList.size();

myList.addFront(1);
myList.printList();
myList.size();
myList.findValue(3);
myList.findValue(4);
myList.printList();
myList.removeAtIndex(2);
myList.printList();

myList.addBack(3);
myList.addBack(5);
myList.addBack(7);
myList.printList();
myList.removeValue(3);
myList.printList();

